<a href="https://colab.research.google.com/github/jazaineam1/Diplomado_ucentral_ME/blob/main/Cuadernos/M1_Clase_3_Manipulaci%C3%B3n_de_datos_en_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## ***Universidad Central***
>## **Dirección de Educación Continua.**



![Imágen1](https://www.ucentral.edu.co/sites/default/files/logo_1.png)

## **Manipulación de datos en Python v2**
## **Módulo 1**

## ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?
En el desarrollo de este caso introduciremos las librerías pandas, numpy y matplotlib para el preprocesamiento, visualización de un conjunto de datos y el desarrollo de algunas medidas estadísticas descriptivas.



##  Introducción

La revisión del comportamiento de los clientes en cualquier empresa es de vital importancia para el desarrollo de estrategias  tanto de mejoramiento de productos como de publicidad y toma de decisiones. Evaluaremos la información obtenida de un conjunto de registro obtenidos de una empresa de telecomunicaciones que cuenta con la revisión de las siguientes variables:

1. totrev: Ingresos del cliente
2. totmou: Total de minutos usados por el cliente
3. area: Área geográfica
4. creditcd: Indicador de tarjeta de crédito
5. eqpdays: Número de días (antigüedad) del equipo actual

Es importante mencionar que la empresa contiene una gran cantidad de clientes y hacer el análisis uno a uno no es lo  óptimo. Con ayuda de la programación y el análisis de datos, se logrará llegar a conclusiones rápidas y acertadas para encontrar relaciones que a simple vista no son tenidas en cuenta.

### ¿Qué se quiere identificar? Teniendo en cuenta la información de la base de datos se quiere identificar si hay una relación entre el total de minutos usados por el cliente con la compañía y el ingreso total de los clientes, es decir, Será posible afirmar que cuando una persona
tiene un mayor ingreso mayor entonces es mayor la utilización de minutos o todo lo contrario?


####  Insumos para tratar este problema


Teniendo en cuenta que la compañía tenia estándares para la entrega de la información, se dispone de un  conjunto de tablas en formato CSV ordenado de la siguiente manera:
La información de este conjunto de datos puede encontrase en [aquí](https://www.kaggle.com/abhinav89/telecom-customer?select=Telecom_customer+churn.csv)

### Objetivos

En este caso usted tendrá que cargar varias bases de datos, hará una exploración básica sobre la información y fusionará las distintas bases para tener una visión general del problema. La idea es que desarrolle las siguientes habilidades:

1. Manejo de la librería pandas de Python para cargar y leer datos;
2. Ideas elementales para una útil transformación de los datos;
3. Construcción y presentación de argumentos válidos que le brindarán una solución para la pregunta expresada más arriba y llegar a una conclusión.
4. Visualización de gráficas elementales que le permitirá entender mejor la información contenida en los datos.

#### Importación de paquetes


Una de las mejores opciones para trabajar con datos tabulares en Python es usar el módulo pandas. La librería `pandas` provee estructuras de datos, genera gráficos de alta calidad con `matplotlib` y se integra de buena forma con otras librerías que usan arrays de `numpy`.

Debemos revisar si la librería se encuentra en nuestro sistema usando el comando

```python
!pip show librería
```

In [ ]:
!pip show pandas
# !pip show numpy
# !pip show matplotlib

Name: pandas
Version: 2.2.2
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: 
Author-email: The Pandas Development Team <pandas-dev@python.org>
License: BSD 3-Clause License

Copyright (c) 2008-2011, AQR Capital Management, LLC, Lambda Foundry, Inc. and PyData Development Team
All rights reserved.

Copyright (c) 2011-2023, Open source contributors.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the copyright holder nor the names of its
  contributors may be u

En el caso de que la librería no se encuentre instalada ejecutamos el comando

```python
!pip install librería
```

Para importar las librerías usamos el comando
```python
import librería as alias
```

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

Luego de que tener los paquetes activados dentro de nuestro ambiente de python podremos usar las fucnciones que se encuentren en la librería.

### Introducción a pandas

Poner información sobre pandas e indicar la diferencia entre series y dataframe

![](https://www.cdn.geeksforgeeks.org/wp-content/uploads/creating_dataframe1.png)

A continuación, mostramos el código a ejecutar para crear una ` Serie`

In [ ]:
Serie1 = pd.Series(
index=[0, 1, 2, 3, 4], name="Variable1", data=["A","B","C","D","E"]
)
Serie1

0    A
1    B
2    C
3    D
4    E
Name: Variable1, dtype: object

In [ ]:
Data_Frame = pd.DataFrame(
index=[0, 1, 2, 3, 4], columns=["Variable2","Variable3"], data=[[1, 1], ["2", 2],[3,3],[4,4],[5,5]]
)
Data_Frame

Variable2  Variable3
0         1          1
1         2          2
2         3          3
3         4          4
4         5          5

Para revisar el tipo de objeto que contiene nuestro DataFrame podemos ejecutar

In [ ]:
Data_Frame.dtypes
#Serie1.dtypes

Variable2    object
Variable3     int64
dtype: object

### Carga de información en `pandas`

Para nuestro  ejercicio usaremos los siguientes conjuntos de datos alojado [acá](https://github.com/jazaineam1/Andes_ADDBD/blob/main/Sesion4/), por favor descarguelo y deje sus datos en el archivo colab:

1. NEW ENGLAND AREA.csv
2. ATLANTIC SOUTH AREA.csv
3. NEW YORK CITY AREA.csv
4. CHICAGO AREA.csv
5. NORTH FLORIDA AREA.csv
6. DC-MARYLAND-VIRGINIA AREA.csv
7. GREAT LAKES AREA.csv
8. NORTHWEST-ROCKY MOUNTAIN AREA.csv
9. MIDWEST AREA.csv
10. SOUTHWEST AREA.csv
11. LOS ANGELES AREA.csv
12. HOUSTON AREA.csv
13. CALIFORNIA NORTH AREA.csv
14. CENTRAL-SOUTH TEXAS AREA.csv
15. DALLAS AREA.csv
16. PHILADELPHIA AREA.csv
17. TENNESSEE AREA.csv
18. OHIO AREA.csv
19. SOUTH FLORIDA AREA.csv

Cada uno de estos conjuntos de datos contiene la información de:

1. *mou_Mean:* Número medio de minutos de uso mensuales
2. *custcare_Mean:* Número medio de llamadas de atención al cliente
3. *area:*  Área geografica
4. *creditcd:* Indicador de tarjeta de crédito
5. *eqpdays:* Número de días (antigüedad) del equipo actual

Además, se incluye _Credito_dias.csv_ con la información de 2 variables adicionales para los registros las cuales son:

1. _creditcd:_ Indicador de tarjeta de crédito
2. _eqpdays:_ Número de días (antigüedad) del equipo actual

Revisaremos inicialmente los datos contenidos en _Credito_dias.csv_ para revisar algunas funciones interesantes de `pandas`, usaremos el comando
```python
pd.read_csv("archivo.csv")
```
Podemos definir el número de registros a imprimir usando
```python
pd.options.display.min_rows=10
````

In [ ]:
!mkdir Bases

mkdir: cannot create directory ‘Bases’: File exists


In [ ]:
!wget https://raw.githubusercontent.com/jazaineam1/Andes_ADDBD/main/Sesion4/Bases/Credito_dias.csv -O Bases/Credito_dias.csv

--2025-03-26 23:19:40--  https://raw.githubusercontent.com/jazaineam1/Andes_ADDBD/main/Sesion4/Bases/Credito_dias.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 283429 (277K) [text/plain]
Saving to: ‘Bases/Credito_dias.csv’

Bases/Credito_dias. 100%[===================>] 276.79K  --.-KB/s    in 0.04s   

2025-03-26 23:19:40 (6.84 MB/s) - ‘Bases/Credito_dias.csv’ saved [283429/283429]



In [ ]:
df=pd.read_csv("Bases/Credito_dias.csv", sep=",",index_col="Customer_ID")
df.head(7)

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
1091652       5441.0   578.80        Y       68
1043846       7294.0  1495.18        Y      460

In [ ]:
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers.readers:

read_csv(filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]', *, sep: 'str | None | lib.NoDefault' = <no_default>, delimiter: 'str | None | lib.NoDefault' = None, header: "int | Sequence[int] | None | Literal['infer']" = 'infer', names: 'Sequence[Hashable] | None | lib.NoDefault' = <no_default>, index_col: 'IndexLabel | Literal[False] | None' = None, usecols: 'UsecolsArgType' = None, dtype: 'DtypeArg | None' = None, engine: 'CSVEngine | None' = None, converters: 'Mapping[Hashable, Callable] | None' = None, true_values: 'list | None' = None, false_values: 'list | None' = None, skipinitialspace: 'bool' = False, skiprows: 'list[int] | int | Callable[[Hashable], bool] | None' = None, skipfooter: 'int' = 0, nrows: 'int | None' = None, na_values: 'Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None' = None, keep_default_na: 'bool' = True, na_filter: 'bool' = True, verbose: 'bool | 

El conjunto de datos se encuentra en el `DataFrame` `df`

sobre el objeto creado podemos usar  algunas funciones  para darnos una idea del comportamiento de la información:
```python
df.head(n)  # imprime los primeros n registros del DataFrame
df.tail(n)  # imprime los primeros n registros del DataFrame
df.shape    # imprime el numero de columnas y filas del DataFrame
df.columns  # imprime el nombre de las columnas del DataFrame
df.index    # imprime el indice de los registros del DataFrame
df.dtypes   # inpreme el tipo de cada una de las columnnas del DataFrame
df.sample(n)# imprime una muestra aleatoria de n registros en el DataFrame
```

In [ ]:
display("primeros 3 registros",df.head(3))
display("últimos registros",df.tail(3))
display(df.columns)
display(df.index)
display("Tipos de registros en el DataFrame",df.dtypes)
display("Muestra akleatoria de 3 registros",df.sample(3))
display("dimensión",df.shape)

'primeros 3 registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675

'últimos registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                   
1096497      3690.0   294.29        Y      212
1087248      3333.0   233.08        Y      290
1099003      5025.0  1436.49        Y     1029

Index(['totmou', 'totrev', 'creditcd', 'eqpdays'], dtype='object')

Index([1082423, 1069232, 1031705, 1089806, 1076695, 1091652, 1043846, 1088967,
       1045239, 1046584,
       ...
       1087145, 1002540, 1044082, 1083458, 1093976, 1027614, 1098422, 1096497,
       1087248, 1099003],
      dtype='int64', name='Customer_ID', length=10000)

'Tipos de registros en el DataFrame'

totmou      float64
totrev      float64
creditcd     object
eqpdays       int64
dtype: object

'Muestra akleatoria de 3 registros'

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1028403        855.0   539.66        Y      642
1045116       4876.0   845.52        N      437
1018390      10260.0  1186.14        Y      695

'dimensión'

(10000, 4)

In [ ]:
df[df["totrev"]<5]

Empty DataFrame
Columns: [totmou, totrev, creditcd, eqpdays]
Index: []

Este conjunto de datos corresponde a 10000 registros y 2 variables

### Manejo de datos duplicados

os dataframes tienen la posibilidad de detectar las filas duplicadas con la función:
```
df.duplicated({columns})
```
Si no especifica {columns}, se busacarán duplicados teniendo en cuenta todas las columnas. Una vez detectados el dtaframe tiene un función para eliminar filas duplicadas, haciendo:

```
df = df.drop_duplicates()
```
También es posible eliminar filas que duplican solo algunos campos. Con el fin de dejar solo una ocurrencia.

```
df.drop_duplicates(['nombre', 'apellido', 'cédula'], keep='last')
```
En este caso, como no todas las colomnas son iguales se conserva solo la última ocurrencia.

### Manejo de datos perdidos o esperados
Son aquellos datos flatantes en la tabla que por alguna razón o error no se encuentran, o parecen con algún valor o etiqueta de no válido.
Para manipular datos perdidos use la explicación dada el siguiente enlace: [DATOS PERDIDOS](http://ligdigonzalez.com/manipulando-datos-perdidos-en-python/)
recuerde que para reemplazar datos perdidos en un dataframe $df$ se usa la función:
```python
promedio = df['totrev'].mean()
df['nota'].replace(np.nan, promedio)
```
en donde la opción 'promedio' busca el valor promedio de la columna (datos numéricos) o la moda (datos categóricos) para reemplazar el valor no válido.

In [ ]:
df.isnull().any(axis=0)

totmou      False
totrev      False
creditcd     True
eqpdays     False
dtype: bool

In [ ]:
df.isnull().any(axis=1)

Customer_ID
1082423    False
1069232    False
1031705    False
1089806    False
1076695    False
           ...  
1027614    False
1098422    False
1096497    False
1087248    False
1099003    False
Length: 10000, dtype: bool

In [ ]:
df["creditcd"].isnull()

Customer_ID
1082423    False
1069232    False
1031705    False
1089806    False
1076695    False
           ...  
1027614    False
1098422    False
1096497    False
1087248    False
1099003    False
Name: creditcd, Length: 10000, dtype: bool

In [ ]:
perdidos = df[(df['creditcd'].isnull()) ]
perdidos

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1072218       5684.0   857.65      NaN      244
1066597        925.0   373.12      NaN      344
1023352      14466.0  2243.76      NaN      387
1045521       2504.0   756.79      NaN      495
1001813      76260.0  5526.11      NaN      361
...              ...      ...      ...      ...
1050795      15591.0  1330.71      NaN      447
1096985       3011.0   522.14      NaN       74
1025203        853.0   650.34      NaN      780
1099258      33660.0  4247.85      NaN      206
1026149       1955.0   606.08      NaN      770

[152 rows x 4 columns]

In [ ]:
promedio = df['totrev'].mean()
df['totrev'].replace(np.nan, promedio)

Customer_ID
1082423     906.58
1069232    1172.25
1031705     577.39
1089806     294.26
1076695     433.16
            ...   
1027614     484.31
1098422     583.36
1096497     294.29
1087248     233.08
1099003    1436.49
Name: totrev, Length: 10000, dtype: float64

También es recomendado usar la funcion `fillna()`

In [ ]:
df['creditcd'].fillna("Y",inplace=True)

<ipython-input-18-3b8336ea03bb>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['creditcd'].fillna("Y",inplace=True)


In [ ]:
df.isnull().any(axis=0)

totmou      False
totrev      False
creditcd    False
eqpdays     False
dtype: bool

In [ ]:
df

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1069232      13552.0  1172.25        N       83
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1076695       4155.0   433.16        N      222
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[10000 rows x 4 columns]



### Cálculo de estadísticas resumen

Las medidas estadísticas resumen son de gran ayuda para comprender como es el comportamiento de la distribución de los datos, los objetos `DataFrame` ofrecen facilidades para hacer el cálculo de algunas estadísticas sobre cada una de las variables.

```python
df["var"].value_counts() # Encuentra la frecuencia de las categorias de una columna (la columna debe ser cualitativa)
df["var"].min()          # Encuentra el mínimo de una variable cuantitativa
df["var"].median()       # Encuentra el mínimo de una variable cuantitativa
df["var"].mean()         # Encuentra el promedio de una variable cuantitativa
df["var"].max()          # Encuentra el máximo de una variable cuantitativa
df["var"].quantile()     # Encuentra los cuantiles de una variable cuantitativa
```


In [ ]:
df["creditcd"].value_counts()

creditcd
Y    6948
N    3052
Name: count, dtype: int64

In [ ]:
df["creditcd"].count()

np.int64(10000)

In [ ]:
df["eqpdays"].min()

-5

La exploración del anterior resultado puede mostrar inconsistencias  en los registros y nos ayudará a tomar decisiones sobre algunos registros, el valor mínimo de la variable `eqpdays` (Número de días (antigüedad) del equipo actual) debería ser estrictamente positivo y estos pueden ser errores de ingreso.

In [ ]:
df["eqpdays"].max()

1823

In [ ]:
df["eqpdays"].median()

338.0

In [ ]:
df["eqpdays"].mean()

np.float64(386.7945)

In [ ]:
df["eqpdays"].quantile(0.25)

np.float64(210.0)

In [ ]:
df["eqpdays"].quantile(0.75)

np.float64(522.0)

Estas estadísticas pueden ser resumidas con el comando:
```python
df["var"].describe()  # Estadísticas resumen principales
```

In [ ]:
df["eqpdays"].describe()

count    10000.000000
mean       386.794500
std        252.720196
min         -5.000000
25%        210.000000
50%        338.000000
75%        522.000000
max       1823.000000
Name: eqpdays, dtype: float64

In [ ]:
df.columns


Index(['totmou', 'totrev', 'creditcd', 'eqpdays'], dtype='object')

In [ ]:
df.dtypes

totmou      float64
totrev      float64
creditcd     object
eqpdays       int64
dtype: object

# Parte 2

### Realizar filtros


| Filtro                      | `df[df[]]`                               | `query()`                                 | Recomendación                                                                                           |
|-----------------------------|------------------------------------------|-------------------------------------------|---------------------------------------------------------------------------------------------------------|
| Igualdad simple             | `df[df['var'] == value]`                 | `df.query("var == value")`                | Ambos métodos funcionan igual de bien.                                                                  |
| Condiciones múltiples (AND) | `df[(cond1) & (cond2)]`                 | `df.query("cond1 and cond2")`             | `query()` tiene una sintaxis más limpia para múltiples condiciones.                                     |
| Condiciones múltiples (OR)  | `df[(cond1) \| (cond2)]`                 | `df.query("cond1 or cond2")`              | `query()` es más legible para múltiples condiciones.                                                    |
| Valores en lista (IN)       | `df[df['var'].isin(list)]`              | `df.query("var in list")`                 | `query()` es más conciso y fácil de leer.                                                               |
| Comparación numérica        | `df[df['var'] > value]`                 | `df.query("var > value")`                 | Ambos métodos son igual de útiles.                                                                      |
| Expresión de texto          | `df[df['var'].str.contains("text")]`    | No soportado                              | `df[df[]]` es obligatorio para expresiones de texto.                                                    |
| Variables externas          | `df[df['var'] > threshold]`             | `df.query("var > @threshold")`            | `query()` es más conveniente con el uso de variables externas al permitir `@variable`.                  |


In [ ]:
df[df['creditcd'].str.contains("N")]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1069232      13552.0  1172.25        N       83
1076695       4155.0   433.16        N      222
1088967       4395.0   668.23        N      333
1097730        810.0   240.94        N      131
1088984      10194.0  1089.04        N      209
...              ...      ...      ...      ...
1053841       2163.0   351.02        N       17
1054709        991.0   482.54        N      460
1002540      82843.0  8927.15        N      182
1044082       6646.0   953.91        N      446
1093976       6531.0   663.21        N      276

[3052 rows x 4 columns]

In [ ]:
df[df['eqpdays'].isin([100,101])]
# df.query("eqpdays in [100,101]")

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1090002       4011.0   391.87        N      101
1068699       5115.0  1049.66        N      100
1065571       4857.0   391.95        Y      100
1077341       6386.0   720.48        Y      100
1075892       9498.0   472.48        Y      101
1060518       3466.0   805.50        Y      100
1059332      11341.0  1009.16        Y      101
1057998       7110.0   664.20        Y      100
1054030       2434.0   637.34        Y      100
1078064       3196.0   636.06        N      101
1011791       2603.0  1073.84        Y      100
1063042      13044.0  1256.11        N      101
1083454      23127.0  1169.64        Y      100
1084679       7419.0   664.94        Y      101
1061421       7144.0   716.91        Y      100
1014026       3926.0  1027.26        Y      101
1053777       2293.0   566.11        N      100

In [ ]:
# df[(cond1) & (cond2)]
df[(df["creditcd"]=="Y") & ( df["eqpdays"]>100)]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1043846       7294.0  1495.18        Y      460
1045239      11544.0  1012.64        Y      498
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[6197 rows x 4 columns]

In [ ]:
df.query("creditcd=='Y' and eqpdays>100")

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1082423       7269.0   906.58        Y      371
1031705       4152.0   577.39        Y      675
1089806       1863.0   294.26        Y      218
1043846       7294.0  1495.18        Y      460
1045239      11544.0  1012.64        Y      498
...              ...      ...      ...      ...
1027614       2570.0   484.31        Y      611
1098422       2105.0   583.36        Y     1243
1096497       3690.0   294.29        Y      212
1087248       3333.0   233.08        Y      290
1099003       5025.0  1436.49        Y     1029

[6197 rows x 4 columns]

In [ ]:
P25 = df['eqpdays'].quantile(0.25)
df[df['eqpdays'] > P25]

totmou   totrev creditcd  eqpdays
Customer_ID                                   
1082423      7269.0   906.58        Y      371
1031705      4152.0   577.39        Y      675
1089806      1863.0   294.26        Y      218
1076695      4155.0   433.16        N      222
1043846      7294.0  1495.18        Y      460
...             ...      ...      ...      ...
1027614      2570.0   484.31        Y      611
1098422      2105.0   583.36        Y     1243
1096497      3690.0   294.29        Y      212
1087248      3333.0   233.08        Y      290
1099003      5025.0  1436.49        Y     1029

[7486 rows x 4 columns]

In [ ]:
!wget https://github.com/jazaineam1/Diplomado_ucentral_ME/raw/main/Datos/agregado_area_total.xlsx -O Bases/agregado_area_total.xlsx

--2025-03-27 00:16:59--  https://github.com/jazaineam1/Diplomado_ucentral_ME/raw/main/Datos/agregado_area_total.xlsx
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/jazaineam1/Diplomado_ucentral_ME/main/Datos/agregado_area_total.xlsx [following]
--2025-03-27 00:16:59--  https://raw.githubusercontent.com/jazaineam1/Diplomado_ucentral_ME/main/Datos/agregado_area_total.xlsx
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 258880 (253K) [application/octet-stream]
Saving to: ‘Bases/agregado_area_total.xlsx’

Bases/agregado_area 100%[===================>] 252.81K  --.-KB/s    in 0.04s   

2025-03-27 00

In [ ]:
agr_df = pd.read_excel("Bases/agregado_area_total.xlsx")

In [ ]:
agr_df

Customer_ID  custcare_Mean new_cell                               area  \
0         1000014       0.000000        Y       CENTRAL-SOUTH_TEXAS_AREA.csv   
1         1000016       1.000000        N                    DALLAS_AREA.csv   
2         1000021       0.333333        U                 TENNESSEE_AREA.csv   
3         1000023       0.000000        N  NORTHWEST-ROCKY_MOUNTAIN_AREA.csv   
4         1000029       0.000000        N               NEW_ENGLAND_AREA.csv   
...           ...            ...      ...                                ...   
9992      1099907      17.333333        U               LOS_ANGELES_AREA.csv   
9993      1099932       0.000000        U               LOS_ANGELES_AREA.csv   
9994      1099963       0.000000        U                 SOUTHWEST_AREA.csv   
9995      1099981       0.333333        U                 SOUTHWEST_AREA.csv   
9996      1099992       0.000000        Y               LOS_ANGELES_AREA.csv   

     creditcd  
0           N  
1           Y  
2           Y  
3           N  
4           Y  
...       ...  
9992        Y  
9993        Y  
9994        Y  
9995      NaN  
9996        Y  

[9997 rows x 5 columns]

La ejecución nos da como resultado una base de datos de 9997 registros correspondiente a la concatenación de todas las áreas. Se debe notar que nuestro conjunto de datos también es indexado por la columna `Customer_ID` y que la dimensión de los registros de está nueva base es menor a la de  `Creditos_dias.csv`.


Cada uno de estos conjuntos de datos contiene la información de:

1. *mou_Mean:* Número medio de minutos de uso mensuales
2. *custcare_Mean:* Número medio de llamadas de atención al cliente
3. *area:*  Área geografica
4. *creditcd:* Indicador de tarjeta de crédito
5. *eqpdays:* Número de días (antigüedad) del equipo actual

In [ ]:
df.columns

Index(['totmou', 'totrev', 'creditcd', 'eqpdays'], dtype='object')

In [ ]:
agr_df.columns

Index(['Customer_ID', 'custcare_Mean', 'new_cell', 'area', 'creditcd'], dtype='object')

 ## Cruce de información

El cruce de información es una operación esencial en el análisis de los datos. Usualmente, la posibilidad de realizar el cruce de dos o más fuentes de información es el primer paso para responder preguntas complejas acerca de los datos. Existen muchas formas de cruzar o combinar tablas de datos. El cruce de estos datos siempre tiene un propósito. Por tanto, se debe realizar un análisis de lo que se requiere antes de iniciar. Para esto, lo primero es conocer bien las fuentes de información que se van a relacionar.

### Uniones (Joins)
Usando la librería pandas se pueden realizar 4 tipos de uniones para cruzar información entre dos fuentes de datos, ver Figura abajo.
* INNER JOIN: este tipo de cruce incluye solo los resgistros que coincieden en las dos tablas
* LEFT JOIN:  en este tipo de cruce se incluyen todos los registros de la primera tabla aunque no aparezcan coincidencias en la segunda. Cuando no hay coincidencias en la segunda tabla los campos correspondientes a esta aparecerán como nulos.
* RIGHT JOIN: en este tipo de cruce se incluyen todos los registros de la segunda tabla aunque no aparezcan coincidencias en la primera. Cuando no hay coincidencias en la primer tabla los campos correspondientes a esta aparecerán como nulos.
* OUTER JOIN O FULL JOIN: combina todos los registros de ambas tablas aunque no existan coincidencias en la otra tabla.

![Los tipos de uniones que se pueden ejecutar en pandas](https://letsdobigdata.files.wordpress.com/2016/03/joins.png)

La combinación a usar depende del análisis que se requiera con los datos. Lo más común es realizar un cruce de tipo Inner Join, debido a que muestra los registros cuando aparecen relacionados en ambas tablas. Sin embargo, en
ocasiones se quiere saber que pasa con todos los registros de la primera tabla aunque no se tengan coincidencias. Alternativamente, quizá, lo menos usual es hacer combinaciones de tipo RIGHT o OUTER. Generalmente, estas son usadas cuando se quiere saber por qué no se tienen coincidencias, o cuáles registros no cumplen con la relación especificada.

Ya que desconocemos cuales son los registros perdidos en los 2 dataframes que hemos construido y es necesario para continuar nuestro preprocesamiento el unir la información.

In [ ]:
df.head(1)

totmou  totrev creditcd  eqpdays
Customer_ID                                  
1082423      7269.0  906.58        Y      371

In [ ]:
agr_df.head(1)

Customer_ID  custcare_Mean new_cell                          area creditcd
0      1000014            0.0        Y  CENTRAL-SOUTH_TEXAS_AREA.csv        N

In [ ]:
agr_df.set_index("Customer_ID",inplace=True)

In [ ]:
agr_df

custcare_Mean new_cell                               area  \
Customer_ID                                                              
1000014           0.000000        Y       CENTRAL-SOUTH_TEXAS_AREA.csv   
1000016           1.000000        N                    DALLAS_AREA.csv   
1000021           0.333333        U                 TENNESSEE_AREA.csv   
1000023           0.000000        N  NORTHWEST-ROCKY_MOUNTAIN_AREA.csv   
1000029           0.000000        N               NEW_ENGLAND_AREA.csv   
...                    ...      ...                                ...   
1099907          17.333333        U               LOS_ANGELES_AREA.csv   
1099932           0.000000        U               LOS_ANGELES_AREA.csv   
1099963           0.000000        U                 SOUTHWEST_AREA.csv   
1099981           0.333333        U                 SOUTHWEST_AREA.csv   
1099992           0.000000        Y               LOS_ANGELES_AREA.csv   

            creditcd  
Customer_ID           
1000014            N  
1000016            Y  
1000021            Y  
1000023            N  
1000029            Y  
...              ...  
1099907            Y  
1099932            Y  
1099963            Y  
1099981          NaN  
1099992            Y  

[9997 rows x 4 columns]

In [ ]:
df3=df.merge(agr_df,left_index=True,right_index=True,how="inner")
df3

totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1082423       7269.0   906.58          Y      371      14.333333        U   
1069232      13552.0  1172.25          N       83       3.666667        U   
1031705       4152.0   577.39          Y      675       3.000000        U   
1089806       1863.0   294.26          Y      218      11.000000        U   
1076695       4155.0   433.16          N      222       0.000000        U   
...              ...      ...        ...      ...            ...      ...   
1027614       2570.0   484.31          Y      611       0.666667        U   
1098422       2105.0   583.36          Y     1243       0.000000        U   
1096497       3690.0   294.29          Y      212       0.666667        Y   
1087248       3333.0   233.08          Y      290       0.000000        U   
1099003       5025.0  1436.49          Y     1029       0.333333        U   

                                  area creditcd_y  
Customer_ID                                        
1082423           NEW_ENGLAND_AREA.csv          Y  
1069232        ATLANTIC_SOUTH_AREA.csv          N  
1031705         NEW_YORK_CITY_AREA.csv          Y  
1089806               CHICAGO_AREA.csv          Y  
1076695         NORTH_FLORIDA_AREA.csv          N  
...                                ...        ...  
1027614               CHICAGO_AREA.csv          Y  
1098422             SOUTHWEST_AREA.csv          Y  
1096497           NEW_ENGLAND_AREA.csv          Y  
1087248      CALIFORNIA_NORTH_AREA.csv          Y  
1099003           LOS_ANGELES_AREA.csv          Y  

[9997 rows x 8 columns]

### Ejercicio 2  
¿Cuáles son los 3 registros que no se incluyen en df3 ?


In [ ]:
df.index

Index([1082423, 1069232, 1031705, 1089806, 1076695, 1091652, 1043846, 1088967,
       1045239, 1046584,
       ...
       1087145, 1002540, 1044082, 1083458, 1093976, 1027614, 1098422, 1096497,
       1087248, 1099003],
      dtype='int64', name='Customer_ID', length=10000)

In [ ]:
df3.index

Index([1082423, 1069232, 1031705, 1089806, 1076695, 1091652, 1043846, 1088967,
       1045239, 1046584,
       ...
       1087145, 1002540, 1044082, 1083458, 1093976, 1027614, 1098422, 1096497,
       1087248, 1099003],
      dtype='int64', name='Customer_ID', length=9997)

In [ ]:
sin_registros=(df.index.isin(df3.index))
sin_registros

array([ True,  True,  True, ...,  True,  True,  True])

In [ ]:
df.iloc[~sin_registros,]

totmou   totrev creditcd  eqpdays
Customer_ID                                    
1050270       1409.0   407.02        Y      102
1020629      15284.0  2370.62        N      445
1020451       3604.0   662.42        Y      675

In [ ]:
agr_df[agr_df.index==1020451]

Empty DataFrame
Columns: [custcare_Mean, new_cell, area, creditcd]
Index: []

Si deseamos realizar una filtración sobre algún valor relevante por ejemplo el área podremos revertir el proceso usando el operador `==`  que retorna `True` en el caso que dos objetos tengan el mismo valor

In [ ]:
new_ingland=df3[df3["area"]=="NEW_ENGLAND_AREA.csv"]
new_ingland

totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1082423       7269.0   906.58          Y      371      14.333333        U   
1076771        288.0   374.98          Y      284       0.000000        U   
1083737       1928.0   396.78          Y      261       0.000000        U   
1078557       2781.0   590.26          Y      276       0.000000        N   
1018015      24069.0  2129.74          Y      199       2.000000        N   
...              ...      ...        ...      ...            ...      ...   
1059454        942.0   625.98          Y      323       0.000000        U   
1036973       2538.0   981.41          Y      688       0.000000        U   
1073829       2812.0   454.64          Y      342       0.000000        U   
1083458      10628.0   965.57          Y      262       6.666667        U   
1096497       3690.0   294.29          Y      212       0.666667        Y   

                             area creditcd_y  
Customer_ID                                   
1082423      NEW_ENGLAND_AREA.csv          Y  
1076771      NEW_ENGLAND_AREA.csv          Y  
1083737      NEW_ENGLAND_AREA.csv          Y  
1078557      NEW_ENGLAND_AREA.csv          Y  
1018015      NEW_ENGLAND_AREA.csv          Y  
...                           ...        ...  
1059454      NEW_ENGLAND_AREA.csv          Y  
1036973      NEW_ENGLAND_AREA.csv          Y  
1073829      NEW_ENGLAND_AREA.csv          Y  
1083458      NEW_ENGLAND_AREA.csv          Y  
1096497      NEW_ENGLAND_AREA.csv          Y  

[525 rows x 8 columns]

El comando   `df3["area"]=="NEW ENGLAND AREA"`

In [ ]:
df3["area"]=="NEW_ENGLAND_AREA.csv"

Customer_ID
1082423     True
1069232    False
1031705    False
1089806    False
1076695    False
           ...  
1027614    False
1098422    False
1096497     True
1087248    False
1099003    False
Name: area, Length: 9997, dtype: bool

devuelve una serie boleana  con el mismo número de filas de la base`df3`, donde cada valor es `True` o `False` dependiendo si el valor de `area` de un registro  toma el valor de "NEW ENGLAND AREA"

## ¿Cuántos registros de df3 pertenecen a  CALIFORNIA_NORTH_AREA.csv?

In [ ]:
california_north=df3.query("area == 'CALIFORNIA_NORTH_AREA.csv'")
california_north.shape

(599, 8)

In [ ]:
california_north

totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1064411       6282.0   729.54          Y      357      15.666667        U   
1029380       1834.0   610.87          Y      575       0.000000        U   
1056368      10517.0  1108.00          N      307       0.666667        Y   
1046285      18750.0  1485.02          Y      534       2.000000        Y   
1010813      19030.0  2275.21          N      198       1.333333        U   
...              ...      ...        ...      ...            ...      ...   
1055596       7515.0   949.41          Y      346       0.666667        U   
1081680       5593.0   631.12          N      313       0.333333        Y   
1032664       2109.0   314.68          Y      564       0.333333        U   
1042602      14584.0  2013.91          Y      523       0.666667        U   
1087248       3333.0   233.08          Y      290       0.000000        U   

                                  area creditcd_y  
Customer_ID                                        
1064411      CALIFORNIA_NORTH_AREA.csv          Y  
1029380      CALIFORNIA_NORTH_AREA.csv          Y  
1056368      CALIFORNIA_NORTH_AREA.csv          N  
1046285      CALIFORNIA_NORTH_AREA.csv          Y  
1010813      CALIFORNIA_NORTH_AREA.csv          N  
...                                ...        ...  
1055596      CALIFORNIA_NORTH_AREA.csv          Y  
1081680      CALIFORNIA_NORTH_AREA.csv          N  
1032664      CALIFORNIA_NORTH_AREA.csv          Y  
1042602      CALIFORNIA_NORTH_AREA.csv          Y  
1087248      CALIFORNIA_NORTH_AREA.csv          Y  

[599 rows x 8 columns]

### Agrupación por área geográfica

En el Analisis de la información se suelen usar funciones de resumen, las cuales plantean el agrupamiento de ciertas características por categorías. Para más información consultar [gropby().](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html).

Para nuestro desarrollo es de interés el revisar si el comportamiento de las variables se mantiene en cada una de las áreas geográficas.


In [ ]:
df3.groupby("area")

In [ ]:
df3.groupby('area')['custcare_Mean'].mean().sort_values(ascending=False)

area
NEW_YORK_CITY_AREA.csv               2.820233
DC-MARYLAND-VIRGINIA_AREA.csv        2.523958
ATLANTIC_SOUTH_AREA.csv              2.099662
GREAT_LAKES_AREA.csv                 2.077869
LOS_ANGELES_AREA.csv                 1.978639
HOUSTON_AREA.csv                     1.895924
PHILADELPHIA_AREA.csv                1.839456
SOUTH_FLORIDA_AREA.csv               1.792350
MIDWEST_AREA.csv                     1.745499
DALLAS_AREA.csv                      1.709781
CENTRAL-SOUTH_TEXAS_AREA.csv         1.696822
CALIFORNIA_NORTH_AREA.csv            1.647190
NORTH_FLORIDA_AREA.csv               1.631222
NEW_ENGLAND_AREA.csv                 1.612063
NORTHWEST-ROCKY_MOUNTAIN_AREA.csv    1.555556
SOUTHWEST_AREA.csv                   1.482877
CHICAGO_AREA.csv                     1.465054
TENNESSEE_AREA.csv                   1.406716
OHIO_AREA.csv                        1.373760
Name: custcare_Mean, dtype: float64

In [ ]:
df3.to_excel("Bases/df3.xlsx")

En `pandas`tenemos una manera más sencilla de mostrar  todas las estadística resumen usando el siguiente comando

In [ ]:
df3[["area","totrev"]].groupby("area").describe()

totrev                                    \
                                    count         mean          std     min   
area                                                                          
ATLANTIC_SOUTH_AREA.csv             592.0   903.497399   697.197572  106.95   
CALIFORNIA_NORTH_AREA.csv           599.0  1136.348013   922.069148  105.14   
CENTRAL-SOUTH_TEXAS_AREA.csv        409.0  1174.366577   920.804942   48.53   
CHICAGO_AREA.csv                    496.0   892.133206   669.068444   59.90   
DALLAS_AREA.csv                     518.0  1258.508784  1267.499852   84.59   
DC-MARYLAND-VIRGINIA_AREA.csv       640.0   941.692188   634.132033   86.78   
GREAT_LAKES_AREA.csv                488.0   987.714324   733.521417  126.56   
HOUSTON_AREA.csv                    458.0  1127.294127   999.987571   92.38   
LOS_ANGELES_AREA.csv                671.0  1031.714382   738.616227   96.74   
MIDWEST_AREA.csv                    685.0  1095.678701   945.251868   90.42   
NEW_ENGLAND_AREA.csv                525.0   941.766038   705.572889  118.51   
NEW_YORK_CITY_AREA.csv             1145.0  1076.263546   869.372842   74.01   
NORTHWEST-ROCKY_MOUNTAIN_AREA.csv   429.0  1083.179091   977.857935   67.80   
NORTH_FLORIDA_AREA.csv              442.0   895.489457   642.729028   37.39   
OHIO_AREA.csv                       437.0  1006.616499   785.713223  107.12   
PHILADELPHIA_AREA.csv               245.0   998.303592   722.348164  133.63   
SOUTHWEST_AREA.csv                  584.0  1036.548630   914.475459   11.10   
SOUTH_FLORIDA_AREA.csv              366.0  1167.546257   891.952255  118.62   
TENNESSEE_AREA.csv                  268.0  1042.700597   886.731980  102.03   

                                                                           
                                        25%      50%        75%       max  
area                                                                       
ATLANTIC_SOUTH_AREA.csv            464.9875  718.370  1135.3050   8171.47  
CALIFORNIA_NORTH_AREA.csv          535.9900  892.340  1396.4550   7511.11  
CENTRAL-SOUTH_TEXAS_AREA.csv       594.3300  895.090  1374.5800   5848.39  
CHICAGO_AREA.csv                   474.0675  718.855  1086.9250   5710.44  
DALLAS_AREA.csv                    581.6975  965.650  1484.8500  13149.85  
DC-MARYLAND-VIRGINIA_AREA.csv      525.7200  787.360  1191.5525   5029.68  
GREAT_LAKES_AREA.csv               503.7325  767.960  1211.8900   5624.71  
HOUSTON_AREA.csv                   530.6400  892.180  1443.1400   8719.53  
LOS_ANGELES_AREA.csv               554.2650  836.770  1271.6150   6120.84  
MIDWEST_AREA.csv                   553.0600  854.350  1338.6900  13426.60  
NEW_ENGLAND_AREA.csv               497.5700  747.130  1185.8000   6939.19  
NEW_YORK_CITY_AREA.csv             562.3400  870.310  1297.9700  10590.01  
NORTHWEST-ROCKY_MOUNTAIN_AREA.csv  524.3700  805.020  1293.2200   7419.73  
NORTH_FLORIDA_AREA.csv             481.5225  758.190  1134.1950   6009.01  
OHIO_AREA.csv                      512.0700  781.040  1230.5600   8007.95  
PHILADELPHIA_AREA.csv              558.7100  849.340  1154.6100   4819.91  
SOUTHWEST_AREA.csv                 508.3275  802.025  1187.3275   8772.57  
SOUTH_FLORIDA_AREA.csv             582.0225  914.600  1451.2550   6954.07  
TENNESSEE_AREA.csv                 486.6800  832.400  1282.6150   6751.70

Estos resultados son iguales a los realizados con el ciclo `for`. Los objetos de tipo `DataFrameGroupBy` permiten una rápida salida de generación de estadísticas para grupos de interés.

### Ejercicio 3

Realice una agrupación por la variable creditcd_x y encuentre el valor de la desviación estándar de totrev

Ver más información [acá](https://pandas.pydata.org/docs/reference/frame.html#computations-descriptive-stats)

In [ ]:
# realice el ejercicio en este espacio

df3[["creditcd_x","totrev"]].groupby('creditcd_x').std()

totrev
creditcd_x            
N           876.554177
Y           848.517680

### Etiquetando valores de total de ingresos en categorías

Revisamos el comportamiento de `totrev` y encontramos una gran variabilidad, en ocasiones la agrupación de variables cuantitativas en diferentes categorías nos ayuda a tener una mejor visión del conjunto de datos.

Crearemos una nueva variable `Nivel_totrev`  que catalogue a la variable total de ingresos `totrev` en 2 posibles niveles, alto en el caso que `totrev`  sea mayor al cuantil 50 y baja en otro caso. cada valor de percentil debe ser calculado por cada `area` para asegurar que el +area geografíca sea evaluada adicionalmente.



In [ ]:
pencentiles50 = df3.groupby("area")["totrev"].quantile(0.5) # percentil 50
pencentiles50

area
ATLANTIC_SOUTH_AREA.csv              718.370
CALIFORNIA_NORTH_AREA.csv            892.340
CENTRAL-SOUTH_TEXAS_AREA.csv         895.090
CHICAGO_AREA.csv                     718.855
DALLAS_AREA.csv                      965.650
DC-MARYLAND-VIRGINIA_AREA.csv        787.360
GREAT_LAKES_AREA.csv                 767.960
HOUSTON_AREA.csv                     892.180
LOS_ANGELES_AREA.csv                 836.770
MIDWEST_AREA.csv                     854.350
NEW_ENGLAND_AREA.csv                 747.130
NEW_YORK_CITY_AREA.csv               870.310
NORTHWEST-ROCKY_MOUNTAIN_AREA.csv    805.020
NORTH_FLORIDA_AREA.csv               758.190
OHIO_AREA.csv                        781.040
PHILADELPHIA_AREA.csv                849.340
SOUTHWEST_AREA.csv                   802.025
SOUTH_FLORIDA_AREA.csv               914.600
TENNESSEE_AREA.csv                   832.400
Name: totrev, dtype: float64

Como nos gustaría etiquetar los ingreos en alto y bajopara  cada área geográfica, haremos uso del método
`np.where()` en la biblioteca numpy. Este método toma una entrada y comprueba una condición lógica: si la
condición es verdadera, devolverá su segundo argumento, mientras que si la condición es falsa, devolverá su
tercer argumento.


In [ ]:
df3=df3.merge(pencentiles50, left_on='area', right_index=True,suffixes=('', '_p50'))

In [ ]:
df3

totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1082423       7269.0   906.58          Y      371      14.333333        U   
1069232      13552.0  1172.25          N       83       3.666667        U   
1031705       4152.0   577.39          Y      675       3.000000        U   
1089806       1863.0   294.26          Y      218      11.000000        U   
1076695       4155.0   433.16          N      222       0.000000        U   
...              ...      ...        ...      ...            ...      ...   
1027614       2570.0   484.31          Y      611       0.666667        U   
1098422       2105.0   583.36          Y     1243       0.000000        U   
1096497       3690.0   294.29          Y      212       0.666667        Y   
1087248       3333.0   233.08          Y      290       0.000000        U   
1099003       5025.0  1436.49          Y     1029       0.333333        U   

                                  area creditcd_y  totrev_p50  
Customer_ID                                                    
1082423           NEW_ENGLAND_AREA.csv          Y     747.130  
1069232        ATLANTIC_SOUTH_AREA.csv          N     718.370  
1031705         NEW_YORK_CITY_AREA.csv          Y     870.310  
1089806               CHICAGO_AREA.csv          Y     718.855  
1076695         NORTH_FLORIDA_AREA.csv          N     758.190  
...                                ...        ...         ...  
1027614               CHICAGO_AREA.csv          Y     718.855  
1098422             SOUTHWEST_AREA.csv          Y     802.025  
1096497           NEW_ENGLAND_AREA.csv          Y     747.130  
1087248      CALIFORNIA_NORTH_AREA.csv          Y     892.340  
1099003           LOS_ANGELES_AREA.csv          Y     836.770  

[9997 rows x 9 columns]

In [ ]:
df_con_etiquetas = df3.copy()
df_con_etiquetas["Nivel_totrev"] = np.where(df_con_etiquetas["totrev"] < df_con_etiquetas["totrev_p50"], "Bajo", "Alto")

In [ ]:
df_con_etiquetas.head()

totmou   totrev creditcd_x  eqpdays  custcare_Mean new_cell  \
Customer_ID                                                                 
1082423       7269.0   906.58          Y      371      14.333333        U   
1069232      13552.0  1172.25          N       83       3.666667        U   
1031705       4152.0   577.39          Y      675       3.000000        U   
1089806       1863.0   294.26          Y      218      11.000000        U   
1076695       4155.0   433.16          N      222       0.000000        U   

                                area creditcd_y  totrev_p50 Nivel_totrev  
Customer_ID                                                               
1082423         NEW_ENGLAND_AREA.csv          Y     747.130         Alto  
1069232      ATLANTIC_SOUTH_AREA.csv          N     718.370         Alto  
1031705       NEW_YORK_CITY_AREA.csv          Y     870.310         Bajo  
1089806             CHICAGO_AREA.csv          Y     718.855         Bajo  
1076695       NORTH_FLORIDA_AREA.csv          N     758.190         Bajo

Ahora podemos hacer una evaluaciónd de como es el comportamiento de los minutos gastados y el nivel de ingresos de los clientes

### ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?

Para explorar la relación entre el nivel de total de ingresos y el número promedio de llamadas, agrupemos por
Nivel_totrev y miremos ingreso total promedio por cada +area geográfica.



In [ ]:
df_con_etiquetas.groupby(['area','Nivel_totrev'])[['totmou']].mean()

totmou
area                              Nivel_totrev              
ATLANTIC_SOUTH_AREA.csv           Alto          10287.102128
                                  Bajo           3421.959459
CALIFORNIA_NORTH_AREA.csv         Alto          11937.216156
                                  Bajo           3177.665552
CENTRAL-SOUTH_TEXAS_AREA.csv      Alto          14306.189935
                                  Bajo           4420.431373
CHICAGO_AREA.csv                  Alto           9733.863737
                                  Bajo           3346.189516
DALLAS_AREA.csv                   Alto          14388.723140
                                  Bajo           4110.544402
DC-MARYLAND-VIRGINIA_AREA.csv     Alto           9822.787500
                                  Bajo           3574.912500
GREAT_LAKES_AREA.csv              Alto          11343.180328
                                  Bajo           3677.053279
HOUSTON_AREA.csv                  Alto          14512.521339
                                  Bajo           4278.943231
LOS_ANGELES_AREA.csv              Alto          11593.681548
                                  Bajo           3461.391045
MIDWEST_AREA.csv                  Alto          11656.333537
                                  Bajo           3378.435673
NEW_ENGLAND_AREA.csv              Alto           9935.876198
                                  Bajo           3384.274809
NEW_YORK_CITY_AREA.csv            Alto          12424.576312
                                  Bajo           4231.835664
NORTHWEST-ROCKY_MOUNTAIN_AREA.csv Alto          12433.277767
                                  Bajo           2999.635514
NORTH_FLORIDA_AREA.csv            Alto           9556.276018
                                  Bajo           3418.746606
OHIO_AREA.csv                     Alto          10047.156423
                                  Bajo           3337.463303
PHILADELPHIA_AREA.csv             Alto          10962.949539
                                  Bajo           3113.770492
SOUTHWEST_AREA.csv                Alto          12293.684977
                                  Bajo           3425.452055
SOUTH_FLORIDA_AREA.csv            Alto          14445.641840
                                  Bajo           4168.125683
TENNESSEE_AREA.csv                Alto          12103.077960
                                  Bajo           3643.611940

## Resumen


1. Leer datos desde archivos.
2. Agregar y manipular datos .
3. Analizar estadísticas de resumen y reunir información.
4. Usar matplotlib para crear gráficos para hacer análisis visual, además de otras librerías.

Para más información

> [https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html)